build/lib/src/generate/find_outputs.dart

```diff
@@ -0,0 +1,28 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+import '../asset/id.dart';
+import '../builder/builder.dart';
+
+/// Collects the expected AssetIds created by [builder] when given [input] based
+/// on the extension configuration.
+Set<AssetId> findOutputs(Builder builder, AssetId input) {
+  var matchingExtensions =
+      builder.buildExtensions.keys.where((e) => input.path.endsWith(e));
+  var outputs = matchingExtensions
+      .expand((e) => _replaceExtensions(input, e, builder.buildExtensions[e]));
+  return new Set<AssetId>.from(outputs);
+}
+
+Iterable<AssetId> _replaceExtensions(
+        AssetId assetId, String oldExtension, List<String> newExtensions) =>
+    newExtensions.map((n) => _replaceExtension(assetId, oldExtension, n));
+
+AssetId _replaceExtension(
+    AssetId assetId, String oldExtension, String newExtension) {
+  var path = assetId.path;
+  return new AssetId(
```

Comment:
Might be worth asserting `path.endsWith(oldExtension)`? And/or add a comment that `assetId.path` must end with `oldExtension`.